### ISSUE TITLE
"Pick" or "Exclude" constructor from class type

### ISSUE THREAD

**Issue by bios21** [Microsoft/TypeScript#29261]
## Search Terms
`intersect`, `type`, `intersect constructor`, `pick new`, `pick constructor`

## Suggestion

```ts
class C1 {
  static S1 = 'foo';
  bar = 'bar';
  constructor(p1, p2, p3) {}
}
type T1 = Pick<typeof C1, 'constructor'>;
// could be also `Pick<typeof C1, new>;`

const obj = {} as T1; // `{}` could be anything
new obj(); // should throw error because of missing params
new obj(null, null, null).bar; // should throw error because `bar` doesn't exist
obj.S1; // should throw error because `S1` doesn't exist
```

## Use Cases
It could be useful in class type intersection and/or for multiple inheritance (like PHP Trait, or Java default methods in interface).
And make mixins more "clean" imho.
The goal is to make a "type mixin" without a "value mixin".

## Examples

For example:
```ts
class C1 {
    public bar: string;
    constructor(p1: string) {}
    public foo() {}
}

class C2 {
    public constructor() {}
    public baz() {}
}

type T1 = typeof C1 & typeof C2;

const a: T1 = null;
a.prototype.foo(); // ok
a.prototype.bar; // ok
a.prototype.baz(); // ok
new a(); // ok (return C2)
new a(''); // ok (return C1)
new a().foo() // error
new a('').baz(); // error

type T2 = typeof C1 & Exclude<typeof C2, new>

const b: T2 = null;
b.prototype.foo(); // ok
b.prototype.bar; // ok
b.prototype.baz(); // ok
new a(); // error <<<
new a(''); // ok (return C1 with C2 prototype)
new a('').foo() // ok (from C1)
new a('').baz(); // ok (from C2) <<<
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [ ] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by weswigham** [comment]
Ahhh so you want to be able to pull the signatures out of a type.

**Comment by bios21** [comment]
Well it could be a way to see it.
My need is only to think of the constructor as a named part of the prototype (like other properties and methods).

But blindly extract a signature out of a function could be a nice feature to!

**Comment by fatcerberus** [comment]
Technically `ClassName.prototype.constructor === ClassName` so it’s already a named method but I don’t think the type system sees it as such.

**Comment by bios21** [comment]
Yeah sure, that's why I don't talk about `constructor` in a "`prototype`" way, but `new` as a method in a "`class` "way (even if, ok, class is only syntactic sugar)

**Comment by fatcerberus** [comment]
To be fair, you did say:
> My need is only to think of the constructor as a named part of the prototype

:wink:

But I know what you meant, no worries.

**Comment by bios21** [comment]
I come from Java 😢 my bad.

**Comment by Kovensky** [comment]
Right now any type mapping completely destroys any call or construct signatures. This is probably related to TypeScript not quite having higher order function types.

You can get at the constructor itself by using `typeof ClassName`, but that also includes all static members and the things that come from `Function.prototype`.

**Comment by sarpik** [comment]
bump - need this.

**Comment by gerarts** [comment]
For those that are looking for a solution to patch / replace a constructor in a class, this works for me.

```ts
interface Cat {}

class Duck {
    /* This should make a Duck instance when constructed */
    
    public static quack: string = "QUACK!";
}

type NotDuck = (new() => Cat) & typeof Duck;

const nonDuck: NotDuck = <NotDuck>Duck;

const soundsLike = nonDuck.quack; // We can still see Duck's static
const whoami = new nonDuck(); // but TypeScript sees this instance as Cat
```

If you want to add Cat to Duck you can also just

```ts
interface Cat {}

class Duck {
    /* This should make a Duck instance when constructed */
}

type NotDuck = (new() => Cat & Duck) & typeof Duck;

const nonDuck: NotDuck = <NotDuck>Duck;

const whoami = new nonDuck(); // TypeScript sees this as Cat & Duck
```
